### ISSUE TITLE
test_distributions.py should seed on setUp, to make it harder to accidentally commit nondeterministic tests

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#4353]
Right now it seeds only if you call `_set_rng_seed` which means it's easy to forget to call it, leading to a nondeterministic test.

Instead, we should modify `setUp` in `common.py` and have it seed Numpy as well as Torch:

```
class TestCase(unittest.TestCase):
    precision = 1e-5
    maxDiff = None

    def setUp(self):
        torch.manual_seed(SEED)
```

CC @fritzo

**Comment by fritzo** [comment]
We should also add some maintainer-instructions for dealing with changing randomness sources in test_distributions.py. In particular there are two types of randomized tests:
1. The easier type of randomized test are tests that should always pass but are initialized with random data. If these fail something is wrong, but it's fine to use a fixed seed.
2. The trickier tests are statistical tests, and these are explicitly marked "this is a randomized test". These have a known positive failure rate (we set `failure_rate=1e-3` by default). We need to balance strength of these tests with annoyance of false alarms. One way that works is to specifically set seeds in each of the randomized tests. When a random generator occasionally changes (as in #4312 ), some of these statistical tests may (rarely) fail. If one fails in this case, it's fine to increment the seed of the failing test (but you shouldn't need to increment it more than once; otherwise something is probably actually wrong).

@ezyang If you add a generic `torch.manual_seed(0)` in a `setUp()`, please do not modify or remove `_set_rng_seed()` calls in any tests marked "this is a randomized test".

**Comment by fritzo** [comment]
I'll try to add a `setUp` method as part of the `OneHotCategorical` PR (the nondeterminism just bit me 😄 )

**Comment by ezyang** [comment]
Thanks @fritzo!

**Comment by apaszke** [comment]
Fixed in #4357
